cmd/src/main.go

```diff
@@ -21,7 +29,21 @@ Usage:
 The options are:
 
 	-config=$HOME/src-config.json    specifies a file containing {"accessToken": "<secret>", "endpoint": "https://sourcegraph.com"}
+      You can use "${}" syntax to reference environment variables, even on Windows,
+      but you may wish to use single-quotes to protect that syntax from shell expansion:
+
+      -config=${UserProfile}/src-config.json or
+      -config='${TMPDIR}/my-config.json'
+
+      [Environment Variables]
```

Comment:
Couple thoughts:
1. It is hard to scan through and see all the args or environment variables. This might be solved with better indentation, formatting, or grouping (e.g. the `[Environment Variables]` looks like a "everything below here is related" heading, when it isn't). What if env variables were defined on the same line as the params?

2. The tests you have added are great! My concern though is that we need tests in the first place. Values can come from the config, a parameter, or an environment variable. Seems overly complex (and the fact that there is no parameter for accessToken feels inconsistent). I don't think you made any unreasonable decisions here because you were just extending what already existed, just the end state makes me feel a bit uncomfortable. Maybe another followup for @slimsag.

Reply:
> My concern though is that we need tests in the first place

Heh, yup, that absolutely crossed my mind. And that underlying sentiment is actually why I said I was prepared for the wontfix -- I can deeply appreciate that this is a polarizing change -- but hey, it makes for great code challenge discussion, right?

But the reasoning breaks out like this:

* testing config permutations
* CI/CD pipelines
* ~OCD is a real thing~

It was turning into a huge PITA to test the sudo tokens work via: edit config, test, edit config, test, ... when there was a super tasty `$SRC_ACCESS_TOKEN` just sitting there taunting me.

Based on my experience reforming the [SourceClear Agent](https://www.sourceclear.com/docs/agent-usage/#environment-variables), the DevSecOps users we interfaced with found the ability to provide hierarchical config values extremely useful. So, the sentiment is not driven entirely by the whims of some random guy on the Internet :-)

And I readily admit just having one env-var was disrupting my sense of harmony.

> and the fact that there is no parameter for accessToken feels inconsistent

It is inconsistent, but purposefully so: it is a grave anti-pattern to even _allow_ users to put access tokens on the command line because ~laziness~ human nature implies that if something is possible, folks will do it and not _think_ about it. Command line credentials leak in two bad places: shell history and `ps -ef`. I am _relatively_ sure that one needs to be either the same uid or root to see env-var values from `ps`.

Reply:
As for point number 1, I am 100% not wedded to any bit of that change other than I really believe strongly in giving users as much "tl;dr" as possible when they ask for `-h`. It has the very pleasing side-effect of meaning the docs stay next to the code, and it is a grave pet peeve of mine for `-h` to say something like "for more information, visit this almost certainly 404 webpage ..."

Reply:
### Parameter for accessToken

I think it is fine that there is no parameter for `accessToken`. As you stated already, it would be dangerous to provide one. I am strongly against providing one (and I suspect Nick is, too).

### Documentation

By and large the documentation additions in this PR are the number one thing I am opposed to. This adds a lot of information that I estimate 70% of people running `src -h` to get information will not use, and that is bad. When I see CLI tools that spew out a lot of information like this, I immediately think "I should Google how to achieve $X instead, it may be quicker"

I suggest one of two options:

1. Remove this documentation and instead just include a single line at the end stating something like the following (which effectively communicates the same information in far less words):

> Flags can also be specified via environment variables, e.g. -config via SRC_CONFIG.

2. Provide a `src -extended-help` flag, which could include this more detailed information. I am not convinced that we will have enough docs to put in this flag without it being (almost) useless, though, so I am far less in favor of this option.

### Should SRC_ACCESS_TOKEN be documented here?

I do not think it needs to be. The reason I say this is because this is (1) documented in the README and (2) we directly link to the authentication instructions in the README if you encounter an authentication error https://github.com/sourcegraph/src-cli/blob/master/cmd/src/api.go#L219-L221

### Testing

Yeah, we don't have any tests and that sucks. The codebase started out very small and is growing quickly, so it is not surprising.

I think the best way to test the entire CLI would be by building the binary (or refactoring the main command to be callable from tests) and then invoking it in tests against sourcegraph.com and verifying the expected output. But this is a separate / distinct task, and there is a lot we can test in isolation in unit tests without these sort of 'end to end' CLI tests.

I think the tests you added here are fine because any test > no test, but I didn't review them. Testing in this codebase is a bigger task to take on as I mentioned above, so I don't want to block this PR on figuring out what an ideal test suite setup looks like.